File: app/src/main/java/eu/darken/bluemusic/util/ui/PreferenceView.java
package eu.darken.bluemusic.util.ui;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.annotation.DrawableRes;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import eu.darken.bluemusic.R;


public class PreferenceView extends RelativeLayout {
    @BindView(R.id.icon) ImageView icon;
    @BindView(R.id.title) TextView title;
    @BindView(R.id.description) TextView description;
    @BindView(R.id.extra) ViewGroup extraContainer;

    public PreferenceView(Context context) {
        this(context, null);
    }

    public PreferenceView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public PreferenceView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        View.inflate(getContext(), R.layout.view_preference, this);
        ButterKnife.bind(this);

        int padding = getResources().getDimensionPixelOffset(R.dimen.default_16dp_padding);
        setPadding(padding, padding, padding, padding);

        TypedArray typedArray = null;
        try {
            typedArray = getContext().getTheme().obtainStyledAttributes(new int[]{android.R.attr.selectableItemBackground});
            setBackground(typedArray.getDrawable(0));
        } finally {
            if (typedArray != null) typedArray.recycle();
            typedArray = null;
        }

        try {
            typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.PreferenceView);
            icon.setImageResource(typedArray.getResourceId(R.styleable.PreferenceView_pvIcon, 0));
            title.setText(typedArray.getResourceId(R.styleable.PreferenceView_pvTitle, 0));

            final int descId = typedArray.getResourceId(R.styleable.PreferenceView_pvDescription, 0);
            if (descId != 0) description.setText(descId);
            else description.setVisibility(GONE);
        } finally {
            if (typedArray != null) typedArray.recycle();
        }
    }

    public void addExtra(View view) {
        extraContainer.addView(view);
        extraContainer.setVisibility(view != null ? VISIBLE : GONE);
    }

    @Override
    public void setEnabled(boolean enabled) {
        super.setEnabled(enabled);
        setEnabledRecursion(this, enabled);
    }

    private static void setEnabledRecursion(ViewGroup vg, boolean enabled) {
        for (int i = 0; i < vg.getChildCount(); i++) {
            View child = vg.getChildAt(i);
            if (child instanceof ViewGroup) setEnabledRecursion((ViewGroup) child, enabled);
            else child.setEnabled(enabled);
        }
    }

    public void setIcon(@DrawableRes int iconRes) {
        icon.setImageResource(iconRes);
    }

    public void setDescription(String desc) {
        description.setText(desc);
    }
}

File: app/src/main/java/eu/darken/bluemusic/util/ui/PreferenceView.kt
package eu.darken.bluemusic.util.ui

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import androidx.annotation.DrawableRes
import eu.darken.bluemusic.R
import eu.darken.bluemusic.databinding.ViewPreferenceBinding

open class PreferenceView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : RelativeLayout(context, attrs, defStyleAttr) {

    private val ui = ViewPreferenceBinding.inflate(LayoutInflater.from(context), this)

    init {
        getContext().theme.obtainStyledAttributes(intArrayOf(android.R.attr.selectableItemBackground)).run {
            background = getDrawable(0)
            recycle()
        }

        getContext().obtainStyledAttributes(attrs, R.styleable.PreferenceView).run {
            ui.icon.setImageResource(getResourceId(R.styleable.PreferenceView_pvIcon, 0))
            ui.title.setText(getResourceId(R.styleable.PreferenceView_pvTitle, 0))
            val descId = getResourceId(R.styleable.PreferenceView_pvDescription, 0)
            if (descId != 0) ui.description.setText(descId) else ui.description.visibility = GONE
            recycle()
        }
    }

    override fun onFinishInflate() {
        val padding = resources.getDimensionPixelOffset(R.dimen.default_16dp_padding)
        setPadding(padding, padding, padding, padding)
        super.onFinishInflate()
    }

    fun addExtra(view: View?) {
        ui.extra.addView(view)
        ui.extra.visibility = if (view != null) VISIBLE else GONE
    }

    override fun setEnabled(enabled: Boolean) {
        super.setEnabled(enabled)
        setEnabledRecursion(this, enabled)
    }

    fun setIcon(@DrawableRes iconRes: Int) {
        ui.icon.setImageResource(iconRes)
    }

    fun setDescription(desc: String?) {
        ui.description.text = desc
    }

    companion object {
        private fun setEnabledRecursion(vg: ViewGroup, enabled: Boolean) {
            for (i in 0 until vg.childCount) {
                val child = vg.getChildAt(i)
                if (child is ViewGroup) setEnabledRecursion(child, enabled) else child.isEnabled = enabled
            }
        }
    }
}